fsspec/parquet.py

```diff
@@ -220,29 +283,81 @@ def _get_parquet_byte_ranges(
             result[path] = {(footer_starts[i], footer_ends[i]): footer_samples[i]}
 
-    # Use cat_ranges to gather the data byte_ranges
-    for i, data in enumerate(fs.cat_ranges(data_paths, data_starts, data_ends)):
-        if data_ends[i] > data_starts[i]:
-            result[data_paths[i]][(data_starts[i], data_ends[i])] = data
+    # Transfer the data byte-ranges into local memory
+    _transfer_ranges(fs, result, data_paths, data_starts, data_ends)
 
     # Add b"PAR1" to header if necessary
     if add_header_magic:
-        for i, path in enumerate(paths):
-            add_magic = True
-            for k in result[path].keys():
-                if k[0] == 0 and k[1] >= 4:
-                    add_magic = False
-                    break
-            if add_magic:
-                result[path][(0, 4)] = b"PAR1"
-
-    # Add b"" for reads beyond end of file
-    for i, path in enumerate(paths):
-        result[path][(file_sizes[i], 2 * file_sizes[i])] = b""
+        _add_header_magic(result)
 
     return result
 
 
+def _get_parquet_byte_ranges_from_metadata(
+    metadata,
+    fs,
+    columns=None,
+    row_groups=None,
+    max_gap=64_000,
+    max_block=256_000_000,
+    engine="auto",
+):
+    """Simplified version of `_get_parquet_byte_ranges` for
+    the case that an engine-specific `metadata` object is
+    provided, and the remote footer metadata does not need to
+    be transfered before calculating the required byte ranges.
+    """
+
+    # Set engine if necessary
+    if isinstance(engine, str):
+        engine = _set_engine(engine)
+
+    # Use "engine" to collect data byte ranges
+    data_paths, data_starts, data_ends = engine._parquet_byte_ranges(
+        columns,
+        row_groups=row_groups,
+        metadata=metadata,
+    )
+
+    # Merge adjacent offset ranges
+    data_paths, data_starts, data_ends = merge_offset_ranges(
+        data_paths,
+        data_starts,
+        data_ends,
+        max_gap=max_gap,
+        max_block=max_block,
+        sort=False,  # Should be sorted
+    )
+
+    # Transfer the data byte-ranges into local memory
+    result = {fn: {} for fn in list(set(data_paths))}
+    _transfer_ranges(fs, result, data_paths, data_starts, data_ends)
+
+    # Add b"PAR1" to header
+    _add_header_magic(result)
+
+    return result
+
+
+def _transfer_ranges(fs, blocks, data_paths, data_starts, data_ends):
+    # Use cat_ranges to gather the data byte_ranges
+    for i, data in enumerate(fs.cat_ranges(data_paths, data_starts, data_ends)):
+        if data_ends[i] > data_starts[i]:
+            blocks[data_paths[i]][(data_starts[i], data_ends[i])] = data
+
+
+def _add_header_magic(data):
+    # Add b"PAR1" to file headers
+    for i, path in enumerate(list(data.keys())):
+        add_magic = True
+        for k in data[path].keys():
+            if k[0] == 0 and k[1] >= 4:
+                add_magic = False
+                break
+        if add_magic:
+            data[path][(0, 4)] = b"PAR1"
```

Comment:
(fastparquet doesn't check this by default)

Reply:
I think it may only be cudf that checks this :/

I do think we need the functionality to make this optional, but I was hoping to attack this later since the overhead is pretty negligible